pki/pki.go

```diff
@@ -0,0 +1,132 @@
+		return "granted with modifications"
+	case StatusRejection:
+		return "rejected"
+	case StatusWaiting:
+		return "the request body part has not yet been processed, expect to hear more later"
+	case StatusRevocationWarning:
+		return "warning: a revocation is imminent"
+	case StatusRevocationNotification:
+		return "a revocation has occurred"
+	default:
+		return "unknown PKIStatus"
+	}
+}
+
+// FailureInfo is PKIFailureInfo defined in RFC 3161 2.4.2.
+type FailureInfo int
+
+const (
+	FailureInfoBadAlg              FailureInfo = 0  // unrecognized or unsupported Algorithm Identifier
+	FailureInfoBadRequest          FailureInfo = 2  // transaction not permitted or supported
+	FailureInfoBadDataFormat       FailureInfo = 5  // the data submitted has the wrong format
+	FailureInfoTimeNotAvailable    FailureInfo = 14 // the TSA's time source is not available
+	FailureInfoUnacceptedPolicy    FailureInfo = 15 // the requested TSA policy is not supported by the TSA.
+	FailureInfoUnacceptedExtension FailureInfo = 16 // the requested extension is not supported by the TSA.
+	FailureInfoAddInfoNotAvailable FailureInfo = 17 // the additional information requested could not be understood or is not available
+	FailureInfoSystemFailure       FailureInfo = 25 // the request cannot be handled due to system failure
+)
+
+// FailureInfos is an array of supported PKIFailureInfo
+var FailureInfos = []FailureInfo{FailureInfoBadAlg, FailureInfoBadRequest, FailureInfoBadDataFormat, FailureInfoTimeNotAvailable,
+	FailureInfoUnacceptedPolicy, FailureInfoUnacceptedExtension, FailureInfoAddInfoNotAvailable, FailureInfoSystemFailure}
+
+// String converts FailureInfo to string
+func (pf FailureInfo) String() string {
+	switch pf {
+	case FailureInfoBadAlg:
+		return "unrecognized or unsupported Algorithm Identifier"
+	case FailureInfoBadRequest:
+		return "transaction not permitted or supported"
+	case FailureInfoBadDataFormat:
+		return "the data submitted has the wrong format"
+	case FailureInfoTimeNotAvailable:
+		return "the TSA's time source is not available"
+	case FailureInfoUnacceptedPolicy:
+		return "the requested TSA policy is not supported by the TSA"
+	case FailureInfoUnacceptedExtension:
+		return "the requested extension is not supported by the TSA"
+	case FailureInfoAddInfoNotAvailable:
+		return "the additional information requested could not be understood or is not available"
+	case FailureInfoSystemFailure:
+		return "the request cannot be handled due to system failure"
+	default:
+		return "unknown PKIFailureInfo"
+	}
+}
+
+// StatusInfo contains status codes and failure information for PKI messages.
+//
+//	PKIStatusInfo ::= SEQUENCE {
+//	 status          PKIStatus,
+//	 statusString    PKIFreeText     OPTIONAL,
+//	 failInfo        PKIFailureInfo  OPTIONAL }
+//
+// PKIStatus        ::= INTEGER
+// PKIFreeText      ::= SEQUENCE SIZE (1..MAX) OF UTF8String
+// PKIFailureInfo   ::= BIT STRING
+//
+// Reference: RFC 3161 2.4.2
+type StatusInfo struct {
```

Comment:
Note for me: We need Validate method for status Info that adheres implementation to https://www.ietf.org/rfc/rfc3161.html#section-2.4.2

```
When the status contains the value zero or one, a TimeStampToken MUST
   be present.  When status contains a value other than zero or one, a
   TimeStampToken MUST NOT be present.  One of the following values MUST
   be contained in status:
```